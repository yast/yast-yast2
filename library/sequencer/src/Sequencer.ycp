/****************************************************************************

Copyright (c) 2002 - 2012 Novell, Inc.
All Rights Reserved.

This program is free software; you can redistribute it and/or
modify it under the terms of version 2 of the GNU General Public License as
published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, contact Novell, Inc.

To contact Novell about this file by physical or electronic mail,
you may find current contact information at www.novell.com

****************************************************************************/
/**
 * File:	modules/Sequencer.ycp
 * Module:	yast2
 * Summary:	Wizard Sequencer
 * Authors:	Michal Svec <msvec@suse.cz>
 * Flags:	Stable
 *
 * $Id$
 *
 * This is an implementation of the wizard sequencer, the tool for
 * processing workflows of dialogs.
 * <br>
 * All errors are reported to y2log, so if anything is unfunctional
 * look to the y2log.
 */

{

module "Sequencer";
textdomain "base";

boolean docheck = true;

/**
 * Test (run) all dialogs in the aliases map
 * @param aliases the map of aliases
 * @return returned values of tested dialogs
 * @see WS documentation for the format of aliases map
 */
define list WS_testall(map aliases) {
    return maplist(any id, any func, aliases, { return eval(func); });
}

/**
 * Check correct types in maps and alias presence for sequence.
 * @param aliases the map of aliases
 * @param sequence the sequence of dialogs
 * @return check passed?
 */
define boolean WS_check(map aliases, map sequence) {
    list<boolean> ret = [];

    /* check if aliases is not a nil */
    if (aliases == nil) {
	y2error(2, "sequencer check: aliases is nil");
	return false;
    }

    /* check if sequence is not a nil */
    if (sequence == nil) {
	y2error(2, "sequencer check: sequence is nil");
	return false;
    }

    /* check if ws_start is in aliases */
    if (aliases["ws_start"]:nil != nil) {
        y2error(2, "sequencer check: ws_start cannot be an alias name");
        ret = add(ret, false);
    }
    else ret = add(ret, true);

    /* check aliases map types */
    list<boolean> ret0 = maplist(any key, any val, aliases, {
        if (!is(key, string)) {
            y2error(2, "sequencer check: not a string: %1", key);
            return false;
        }
        else if (is(val, list)) {
            if(size((list) val) < 2) {
                y2error(2, "sequencer check: list size too small: %1 (key: %2)", size((list) val), key);
                return false;
            }
/* FIXME: use function pointers
            else if (!is(select((list) val, 0, nil), term)) {
                y2error(2, "sequencer check: not a term: %1", select((list) val, 0, nil));
                return false;
            }
*/
            else if (!is(select((list) val, 1, nil), boolean)) {
                y2error(2, "sequencer check: not a boolean: %1", select((list) val, 1, nil));
                return false;
            }
            else return true;
        }
/* FIXME: use function pointers
        else if (!is(val, term)) {
            y2error(2, "sequencer check: not a term: %1", val);
            return false;
        }
*/
	else return true;
    });
    ret = flatten([ret, ret0]);

    /* check if ws_start is in sequence */
    if (sequence["ws_start"]:nil == nil) {
        y2error(2, "sequencer check: ws_start needs to be defined");
        ret = add(ret, false);
    }
    else ret = add(ret, true);

    /* check all aliases in sequence */
    ret0 = maplist(any key, any val, sequence, {
        if (key=="ws_start") {
            if (!is(val, symbol) && aliases[val]:nil == nil) {
                y2error(2, "sequencer check: alias not found: %1", val);
                return false;
            }
            else return true;
        }
        else if (aliases[key]:nil == nil) {
            y2error(2, "sequencer check: alias not found: %1", key);
            return false;
        }
        else if (!is(val, map)) {
            y2error(2, "sequencer check: not a map: %1 %2", key, val);
            return false;
        }
        else {
            list<boolean> ret1 = maplist(any k, any v, (map) val, {
                if (!is(k, symbol)) {
                    y2error(2, "sequencer check: not a symbol: %1", k);
                    return false;
                }
                else if (!is(v, symbol) && aliases[v]:nil == nil) {
                    y2error(2, "sequencer check: alias not found: %1", v);
                    return false;
                }
                else return true;
            });
            if (find(boolean n, ret1, { return n == false; }) != nil) return false;
            return true;
        }
    });
    ret = flatten([ret, ret0]);

    /* check that all aliases are used */
    ret0 = maplist(any key, any val, aliases, {
        if (!haskey(sequence, key)) {
            y2warning(2, "sequencer check: alias not used: %1", key);
            // return false;
        }
	return true;
    });
    ret = flatten([ret, ret0]);

    if (find(boolean n, ret, { return n == false; }) != nil) return false;
    return true;
}

/**
 * Report error and return nil
 * @param error the error message text
 * @return always nil
 * @see bug #6474
 */
define any WS_error(string error) {
    y2error(1, "sequencer: %1", error);
    return nil;
}

/**
 * Find an aliases in the aliases map
 * @param aliases map of aliases
 * @param alias given alias
 * @return term belonging to the given alias or nil, if error
 */
define any WS_alias(map aliases, string alias) {
    any found = aliases[alias]:nil;
    if (found == nil)
        return WS_error(sformat("Alias not found: %1", alias));
    if (is(found, list)) {
        if (size((list) found) <= 0)
	    return WS_error(sformat("Invalid alias: %1", found));
	found = select((list) found, 0, nil);
    }
    if (found == nil)
	return WS_error(sformat("Invalid alias: %1", found));
/* FIXME: use function pointers
    if (is(found, term)) */
        return found;
    return WS_error(sformat("Invalid alias: %1", found));
}

/**
 * Decide if an alias is special
 * @param aliases map of aliases
 * @param alias given alias
 * @return true if the given alias is special or nil, if not found
 */
define boolean WS_special(map aliases, string alias) {
    any found = aliases[alias]:nil;
    if (found == nil)
        return (boolean) WS_error(sformat("Alias not found: %1", alias));
    boolean ret = false;
    if (is(found, list))
        if (size((list) found) > 1)
            ret = (boolean) select((list) found, 1, nil);
    return ret;
}

/**
 * Find a next item in the sequence
 * @param sequence sequence of dialogs
 * @param current current dialog
 * @param ret returned value (determines the next dialog)
 * @return next dialog (symbol), WS action (string) or nil, if error (current or next not found)
 */
define any WS_next(map sequence, string current, symbol ret) {
    map found = (map)(sequence[current]:nil);
    if (found == nil) return WS_error(sformat("Current not found: %1", current));
    /* string|symbol next */
    any next = found[ret]:nil;
    if (next == nil) return WS_error(sformat("Symbol not found: %1", ret));
    return next;
}

/**
 * Run a function from the aliases map
 * @param aliases map of aliases
 * @param id function to run
 * @return returned value from function or nil, if function is nil or returned something else than symbol
 */
define symbol WS_run(map aliases, string id) {
    y2debug("Running: %1", id);
    any function = nil;

    function = WS_alias(aliases, id);
    if (function == nil)
        return (symbol) WS_error(sformat("Bad id: %1", id));

    any ret = eval(function);

    if (!is(ret, symbol))
	return (symbol) WS_error(sformat("Returned value not symbol: %1", ret));

    return (symbol) ret;
}

/**
 * Push one item to the stack
 * @param stack stack of previously run dialogs
 * @param item item to be pushed
 * @return the new stack or nil, if the stack is nil
 */
define list WS_push(list stack, any item) {
    if (stack == nil)
        return nil;

    if (!contains(stack, item))
        return add(stack, item);

    boolean found = false;
    list newstack = filter(any v, stack, {
        if (found) return false;
        if (v == item) found=true;
        return true;
    });

    return newstack;
}

/**
 * Pop one item from the stack (remove an item and return the stack top item)
 * @param stack stack of previously run dialogsk
 * @return [ new stack, poped value ] or nil if the stack is empty or nil
 */
define list WS_pop(list stack) {
    if (stack == nil) return nil;
    integer num = size(stack);
    if (num < 2) return nil;
    list newstack = remove(stack, num-1);
    any poped = select(stack, num-2, nil);
    return [ newstack, poped ];
}

/**
 * The Wizard Sequencer
 * @param aliases the map of aliases
 * @param sequence the sequence of dialogs
 * @return final symbol or nil, if error (see the y2log)
 */
global define symbol Run(map aliases, map sequence) {

    /* Check aliases and sequence correctness */
    if (docheck && WS_check(aliases, sequence) != true)
	return (symbol) WS_error("CHECK FAILED");

    list stack = [];
    /* string|symbol current */
    any current = sequence["ws_start"]:nil;
    if (current == nil)
	return (symbol) WS_error("Starting dialog not found");

    while (true) {
	if (is(current, symbol)) {
	    y2debug("Finished");
	    return (symbol) current;
	}

	stack = WS_push(stack, current);
	y2debug("stack=%1", stack);
	any ret = WS_run(aliases, (string) current);

	if (ret == nil || !is(ret, symbol))
	    return (symbol) WS_error(sformat("Invalid ret: %1", ret));

	else if (ret == `back) {
	    y2debug("Back");
	    list poped = [];
	    boolean special = true;
	    do {
		if (size(stack)<2) return `back;
		poped = WS_pop(stack);
		y2debug("poped=%1", poped);
		current = select(poped, 1, nil);
		stack = (list) select(poped, 0, nil);
		special = WS_special(aliases, (string) current);
		y2debug("special=%1", special);
	    } while (special);
	}

	else {
	    y2debug("ret=%1", ret);
	    current = WS_next(sequence, (string) current, (symbol) ret);
	    y2debug("current=%1", current);
	    if (current == nil)
		return (symbol) WS_error(sformat("Next not found: %1", current));
	}
    }

    /* Not reached */
    return nil;
}

/* EOF */
}
