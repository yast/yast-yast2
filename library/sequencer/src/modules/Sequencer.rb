# encoding: utf-8

# ***************************************************************************
#
# Copyright (c) 2002 - 2012 Novell, Inc.
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of version 2 of the GNU General Public License as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, contact Novell, Inc.
#
# To contact Novell about this file by physical or electronic mail,
# you may find current contact information at www.novell.com
#
# ***************************************************************************
# File:	modules/Sequencer.ycp
# Module:	yast2
# Summary:	Wizard Sequencer
# Authors:	Michal Svec <msvec@suse.cz>
# Flags:	Stable
#
# $Id$
#
# This is an implementation of the wizard sequencer, the tool for
# processing workflows of dialogs.
# <br>
# All errors are reported to y2log, so if anything is unfunctional
# look to the y2log.
require "yast"

module Yast
  class SequencerClass < Module
    def main
      textdomain "base"

      @docheck = true
    end

    # Test (run) all dialogs in the aliases map
    # @param [Hash] aliases the map of aliases
    # @return returned values of tested dialogs
    # @see WS documentation for the format of aliases map
    def WS_testall(aliases)
      aliases = deep_copy(aliases)
      Builtins.maplist(aliases) { |_id, func| Builtins.eval(func) }
    end

    # Check correct types in maps and alias presence for sequence.
    # @param [Hash] aliases the map of aliases
    # @param [Hash] sequence the sequence of dialogs
    # @return check passed?
    def WS_check(aliases, sequence)
      aliases = deep_copy(aliases)
      sequence = deep_copy(sequence)
      ret = []

      # check if aliases is not a nil
      if aliases.nil?
        Builtins.y2error(2, "sequencer check: aliases is nil")
        return false
      end

      # check if sequence is not a nil
      if sequence.nil?
        Builtins.y2error(2, "sequencer check: sequence is nil")
        return false
      end

      # check if ws_start is in aliases
      if Ops.get(aliases, "ws_start")
        Builtins.y2error(2, "sequencer check: ws_start cannot be an alias name")
        ret = Builtins.add(ret, false)
      else
        ret = Builtins.add(ret, true)
      end

      # check aliases map types
      ret0 = Builtins.maplist(aliases) do |key, val|
        if !Ops.is_string?(key)
          Builtins.y2error(2, "sequencer check: not a string: %1", key)
          next false
        elsif Ops.is_list?(val)
          if Ops.less_than(Builtins.size(Convert.to_list(val)), 2)
            Builtins.y2error(
              2,
              "sequencer check: list size too small: %1 (key: %2)",
              Builtins.size(Convert.to_list(val)),
              key
            )
            next false
          # FIXME: use function pointers
          #             else if (!is(select((list) val, 0, nil), term)) {
          #                 y2error(2, "sequencer check: not a term: %1", select((list) val, 0, nil));
          #                 return false;
          #             }
          elsif !Ops.is_boolean?(Ops.get(Convert.to_list(val), 1))
            Builtins.y2error(
              2,
              "sequencer check: not a boolean: %1",
              Ops.get(Convert.to_list(val), 1)
            )
            next false
          else
            next true
          end
        else
          # FIXME: use function pointers
          #         else if (!is(val, term)) {
          #             y2error(2, "sequencer check: not a term: %1", val);
          #             return false;
          #         }
          next true
        end
      end
      ret = Builtins.flatten([ret, ret0])

      # check if ws_start is in sequence
      if Ops.get(sequence, "ws_start").nil?
        Builtins.y2error(2, "sequencer check: ws_start needs to be defined")
        ret = Builtins.add(ret, false)
      else
        ret = Builtins.add(ret, true)
      end

      # check all aliases in sequence
      ret0 = Builtins.maplist(sequence) do |key, val|
        if key == "ws_start"
          next true if Ops.is_symbol?(val) || !Ops.get(aliases, val).nil?

          Builtins.y2error(2, "sequencer check: alias not found: %1", val)
          false
        elsif Ops.get(aliases, key).nil?
          Builtins.y2error(2, "sequencer check: alias not found: %1", key)
          false
        elsif !Ops.is_map?(val)
          Builtins.y2error(2, "sequencer check: not a map: %1 %2", key, val)
          false
        else
          ret1 = Builtins.maplist(Convert.to_map(val)) do |k, v|
            if !Ops.is_symbol?(k)
              Builtins.y2error(2, "sequencer check: not a symbol: %1", k)
              false
            elsif !Ops.is_symbol?(v) && Ops.get(aliases, v).nil?
              Builtins.y2error(2, "sequencer check: alias not found: %1", v)
              false
            else
              true
            end
          end
          ret1.all?
        end
      end
      ret = Builtins.flatten([ret, ret0])

      # check that all aliases are used
      ret0 = Builtins.maplist(aliases) do |key, _val|
        if !Builtins.haskey(sequence, key)
          Builtins.y2warning(2, "sequencer check: alias not used: %1", key)
          # return false;
        end
        true
      end
      ret = Builtins.flatten([ret, ret0])

      ret.all? { |v| v }
    end

    # Report error and return nil
    # @param [String] error the error message text
    # @return always nil
    # @see bug #6474
    def WS_error(error)
      Builtins.y2error(1, "sequencer: %1", error)
      nil
    end

    # Find an aliases in the aliases map
    # @param [Hash] aliases map of aliases
    # @param [String] alias given alias
    # @return [Yast::Term] belonging to the given alias or nil, if error
    def WS_alias(aliases, alias_)
      aliases = deep_copy(aliases)
      found = Ops.get(aliases, alias_)
      if found.nil?
        return WS_error(Builtins.sformat("Alias not found: %1", alias_))
      end
      if Ops.is_list?(found)
        if Ops.less_or_equal(Builtins.size(Convert.to_list(found)), 0)
          return WS_error(Builtins.sformat("Invalid alias: %1", found))
        end
        found = Ops.get(Convert.to_list(found), 0)
      end
      if found.nil?
        return WS_error(Builtins.sformat("Invalid alias: %1", found))
      end

      deep_copy(found)
    end

    # Decide if an alias is special
    # @param [Hash] aliases map of aliases
    # @param [String] alias given alias
    # @return true if the given alias is special or nil, if not found
    def WS_special(aliases, alias_)
      aliases = deep_copy(aliases)
      found = Ops.get(aliases, alias_)
      if found.nil?
        return Convert.to_boolean(
          WS_error(Builtins.sformat("Alias not found: %1", alias_))
        )
      end
      ret = false
      if Ops.is_list?(found)
        if Ops.greater_than(Builtins.size(Convert.to_list(found)), 1)
          ret = Ops.get_boolean(Convert.to_list(found), 1)
        end
      end
      ret
    end

    # Find a next item in the sequence
    # @param [Hash] sequence sequence of dialogs
    # @param [String] current current dialog
    # @param [Symbol] ret returned value (determines the next dialog)
    # @return next dialog (symbol), WS action (string) or nil, if error (current or next not found)
    def WS_next(sequence, current, ret)
      sequence = deep_copy(sequence)
      found = Ops.get_map(sequence, current)
      if found.nil?
        return WS_error(Builtins.sformat("Current not found: %1", current))
      end
      # string|symbol next
      next_ = Ops.get(found, ret)
      if next_.nil?
        return WS_error(Builtins.sformat("Symbol not found: %1", ret))
      end
      deep_copy(next_)
    end

    # Run a function from the aliases map
    # @param [Hash] aliases map of aliases
    # @param [String] id function to run
    # @return returned value from function or nil, if function is nil or returned something else than symbol
    def WS_run(aliases, id)
      aliases = deep_copy(aliases)
      Builtins.y2debug("Running: %1", id)

      function = WS_alias(aliases, id)
      if function.nil?
        return Convert.to_symbol(WS_error(Builtins.sformat("Bad id: %1", id)))
      end

      ret = Builtins.eval(function)

      if !Ops.is_symbol?(ret)
        return Convert.to_symbol(
          WS_error(Builtins.sformat("Returned value not symbol: %1", ret))
        )
      end

      Convert.to_symbol(ret)
    end

    # Push one item to the stack
    # @param [Array] stack stack of previously run dialogs
    # @param [Object] item item to be pushed
    # @return the new stack or nil, if the stack is nil
    def WS_push(stack, item)
      stack = deep_copy(stack)
      item = deep_copy(item)
      return nil if stack.nil?

      return Builtins.add(stack, item) if !Builtins.contains(stack, item)

      found = false
      newstack = Builtins.filter(stack) do |v|
        next false if found
        found = true if v == item
        true
      end

      deep_copy(newstack)
    end

    # Pop one item from the stack (remove an item and return the stack top item)
    # @param [Array] stack stack of previously run dialogsk
    # @return [ new stack, poped value ] or nil if the stack is empty or nil
    def WS_pop(stack)
      stack = deep_copy(stack)
      return nil if stack.nil?
      num = Builtins.size(stack)
      return nil if Ops.less_than(num, 2)
      newstack = Builtins.remove(stack, Ops.subtract(num, 1))
      poped = Ops.get(stack, Ops.subtract(num, 2))
      [newstack, poped]
    end

    # The Wizard Sequencer
    # @param [Hash] aliases the map of aliases
    # @param [Hash] sequence the sequence of dialogs
    # @return final symbol or nil, if error (see the y2log)
    def Run(aliases, sequence)
      aliases = deep_copy(aliases)
      sequence = deep_copy(sequence)
      # Check aliases and sequence correctness
      if @docheck && WS_check(aliases, sequence) != true
        return Convert.to_symbol(WS_error("CHECK FAILED"))
      end

      stack = []
      # string|symbol current
      current = Ops.get(sequence, "ws_start")
      if current.nil?
        return Convert.to_symbol(WS_error("Starting dialog not found"))
      end

      loop do
        if Ops.is_symbol?(current)
          Builtins.y2debug("Finished")
          return Convert.to_symbol(current)
        end

        stack = WS_push(stack, current)
        Builtins.y2debug("stack=%1", stack)
        ret = WS_run(aliases, Convert.to_string(current))

        if ret.nil? || !Ops.is_symbol?(ret)
          return Convert.to_symbol(
            WS_error(Builtins.sformat("Invalid ret: %1", ret))
          )
        elsif ret == :back
          Builtins.y2debug("Back")
          poped = []
          special = true
          loop do
            return :back if Ops.less_than(Builtins.size(stack), 2)
            poped = WS_pop(stack)
            Builtins.y2debug("poped=%1", poped)
            current = Ops.get(poped, 1)
            stack = Ops.get_list(poped, 0)
            special = WS_special(aliases, Convert.to_string(current))
            Builtins.y2debug("special=%1", special)
            break if !special
          end
        else
          Builtins.y2debug("ret=%1", ret)
          current = WS_next(
            sequence,
            Convert.to_string(current),
            Convert.to_symbol(ret)
          )
          Builtins.y2debug("current=%1", current)
          if current.nil?
            return Convert.to_symbol(
              WS_error(Builtins.sformat("Next not found: %1", current))
            )
          end
        end
      end

      # Not reached
      nil
    end

    publish variable: :docheck, type: "boolean", private: true
    publish function: :WS_testall, type: "list (map)", private: true
    publish function: :WS_check, type: "boolean (map, map)", private: true
    publish function: :WS_error, type: "any (string)", private: true
    publish function: :WS_alias, type: "any (map, string)", private: true
    publish function: :WS_special, type: "boolean (map, string)", private: true
    publish function: :WS_next, type: "any (map, string, symbol)", private: true
    publish function: :WS_run, type: "symbol (map, string)", private: true
    publish function: :WS_push, type: "list (list, any)", private: true
    publish function: :WS_pop, type: "list (list)", private: true
    publish function: :Run, type: "symbol (map, map)"
  end

  Sequencer = SequencerClass.new
  Sequencer.main
end
