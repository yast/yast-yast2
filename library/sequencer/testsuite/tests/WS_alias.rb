# encoding: utf-8

# ***************************************************************************
#
# Copyright (c) 2002 - 2012 Novell, Inc.
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of version 2 of the GNU General Public License as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, contact Novell, Inc.
#
# To contact Novell about this file by physical or electronic mail,
# you may find current contact information at www.novell.com
#
# ***************************************************************************
module Yast
  class WSAliasClient < Client
    def main
      Yast.include self, "testsuite.rb"
      Yast.import "Sequencer"

      @aliases = {
        "1" => nil,
        "2" => ->() { func1 },
        "3" => [],
        "4" => [nil],
        "5" => [->() { func2 }],
        "6" => [->() { func3 }, :blah]
      }

      TEST(->() { Sequencer.WS_alias({}, "blah") }, [], nil)
      TEST(->() { Sequencer.WS_alias(@aliases, "blah") }, [], nil)
      TEST(->() { Sequencer.WS_alias(@aliases, "1") }, [], nil)
      TEST(->() { Builtins.eval(Sequencer.WS_alias(@aliases, "2")) }, [], nil)
      TEST(->() { Sequencer.WS_alias(@aliases, "3") }, [], nil)
      TEST(->() { Sequencer.WS_alias(@aliases, "4") }, [], nil)
      TEST(->() { Builtins.eval(Sequencer.WS_alias(@aliases, "5")) }, [], nil)
      TEST(->() { Builtins.eval(Sequencer.WS_alias(@aliases, "6")) }, [], nil)

      nil
    end

    def func1
      :func1
    end

    def func2
      :func2
    end

    def func3
      :func3
    end
  end
end

Yast::WSAliasClient.new.main
