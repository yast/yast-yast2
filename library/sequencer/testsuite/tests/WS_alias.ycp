/****************************************************************************

Copyright (c) 2002 - 2012 Novell, Inc.
All Rights Reserved.

This program is free software; you can redistribute it and/or
modify it under the terms of version 2 of the GNU General Public License as
published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, contact Novell, Inc.

To contact Novell about this file by physical or electronic mail,
you may find current contact information at www.novell.com

****************************************************************************/
{

include "testsuite.ycp";
import "Sequencer";

symbol func1() { return `func1; }
symbol func2() { return `func2; }
symbol func3() { return `func3; }

map aliases = $[
    "1" : nil,
    "2" : ``(func1()),
    "3" : [],
    "4" : [ nil ],
    "5" : [ ``(func2()) ],
    "6" : [ ``(func3()), `blah ]
];

TEST(``(Sequencer::WS_alias($[], "blah")), [], nil);
TEST(``(Sequencer::WS_alias(aliases, "blah")), [], nil);
TEST(``(Sequencer::WS_alias(aliases, "1")), [], nil);
TEST(``(eval(Sequencer::WS_alias(aliases, "2"))), [], nil);
TEST(``(Sequencer::WS_alias(aliases, "3")), [], nil);
TEST(``(Sequencer::WS_alias(aliases, "4")), [], nil);
TEST(``(eval(Sequencer::WS_alias(aliases, "5"))), [], nil);
TEST(``(eval(Sequencer::WS_alias(aliases, "6"))), [], nil);

}
