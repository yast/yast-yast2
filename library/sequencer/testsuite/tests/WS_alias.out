Log	sequencer: Alias not found: blah
Return	nil
Log	sequencer: Alias not found: blah
Return	nil
Log	sequencer: Alias not found: 1
Return	nil
Return	`func1
Log	sequencer: Invalid alias: []
Return	nil
Log	sequencer: Invalid alias: nil
Return	nil
Return	`func2
Return	`func3
